crates/bevy_ecs/src/system/function_system.rs

```diff
@@ -156,6 +156,7 @@ impl<Param: SystemParam> SystemState<Param> {
         &'s mut self,
         world: &'w World,
     ) -> <Param::Fetch as SystemParamFetch<'w, 's>>::Item {
+        // TODO: Only increment if fetch is not read-only, otherwise just read the value.
         let change_tick = world.increment_change_tick();
```

Comment:
Could we check if the type id of the Fetch and ReadonlyFetch associated types match? Or maybe check their Access?

Reply:
@TheRawMeatball suggested using an associated const like this one.

https://github.com/bevyengine/bevy/blob/d8974e7c3d84189313268f475ceeff9393cedffa/crates/bevy_ecs/src/query/fetch.rs#L69-L74

Then the compiler could optimize away any `if` statements.

Reply:
Right, this makes sense. IMO this is probably worth doing (to save on work, not to extend the already absurd lifespan), especially if we can reuse that const for read-only run criteria.

Reply:
Yeah, if no complaints, I'll leave that for another PR.